pytorch_lightning/overrides/data_parallel.py

```diff
@@ -151,74 +152,98 @@ def parallel_apply(self, replicas, inputs, kwargs):
+#         self._sync_params()
+#         self.reducer_reset_hooks()
+#         fx_called: str = ''
+#
+#         if self.device_ids:
+#
+#             inputs, kwargs = self.scatter(inputs, kwargs, self.device_ids)
+#             if len(self.device_ids) == 1:
+#                 # --------------
+#                 # LIGHTNING MOD
+#                 # --------------
+#                 # normal
+#                 # output = self.module(*inputs[0], **kwargs[0])
+#                 # lightning
+#                 if self.module.training:
+#                     output = self.module.training_step(*inputs[0], **kwargs[0])
+#                     fx_called = 'training_step'
+#                 elif self.module.testing:
+#                     output = self.module.test_step(*inputs[0], **kwargs[0])
+#                     fx_called = 'test_step'
+#                 else:
+#                     output = self.module.validation_step(*inputs[0], **kwargs[0])
+#                     fx_called = 'validation_step'
+#             else:
+#                 outputs = self.parallel_apply(self._module_copies[:len(inputs)], inputs, kwargs)
+#                 output = self.gather(outputs, self.output_device)
+#         else:
+#             # output = self.module(*inputs, **kwargs)
+#             # normal lightning (ddp_cpu)
+#             if self.module.training:
+#                 output = self.module.training_step(*inputs, **kwargs)
+#             elif self.module.testing:
+#                 output = self.module.test_step(*inputs, **kwargs)
+#             else:
+#                 output = self.module.validation_step(*inputs, **kwargs)
+#
+#         if not self._reducer_prepared_for_backwards and self.PREPARE_FOR_BACKWARDS:
+#             self.reducer_prepare_for_backwards(output)
+#
+#         if output is None:
+#             warn_missing_output(f'{fx_called} returned None. Did you forget to return an output')
+#         return output
+#
+#     def reducer_prepare_for_backwards(self, output):
+#         self._reducer_prepared_for_backwards = True
+#         if torch.is_grad_enabled():
+#             # We'll return the output object verbatim since it is a freeform
+#             # object. We need to find any tensors in this object, though,
+#             # because we need to figure out which parameters were used during
+#             # this forward pass, to ensure we short circuit reduction for any
+#             # unused parameters. Only if `find_unused_parameters` is set.
+#             if self.find_unused_parameters:
+#                 self.reducer.prepare_for_backward(list(_find_tensors(output)))
+#             else:
+#                 self.reducer.prepare_for_backward([])
+#
+#     def reducer_reset_hooks(self):
+#         self._reducer_prepared_for_backwards = False
 
-        if output is None:
-            warn_missing_output(f'{fx_called} returned None. Did you forget to return an output')
-        return output
 
-    def reducer_prepare_for_backwards(self, output):
```

Comment:
I'm not sure where this should go, it requires the reducer from ddp

Reply:
maybe a hook?
https://pytorch.org/docs/stable/generated/torch.nn.Flatten.html?highlight=hook#torch.nn.Flatten.register_forward_pre_hook

Reply:
@SeanNaren do you remember the conversation we started in #4976 about this? You had an idea there, I'm trying to understand it, maybe you can explain again? :)

Reply:
I have been thinking about this one maybe too much ahah, but I didn't find a better way to do it as `backward` and `optimizer.step` are being called in `training_step` and `DDP reducer` is being called on training_step output

Reply:
could open a PR in PyTorch to at least move it to a function we can use